Boost.multiprecision

ivme opened the pull request:
Adds Boost.multiprecision as an optional replacement for GMP.
Specify -D INTEGER_CLASS=boostmp in cmake configuration.
Boost.multiprecision must be installed on the system separately.  Testing was done with Boost 1.62.0.
This build removes SymEngine's dependence on GMP.  With this build, all GPL/LGPL code is removed by the preprocessor prior to compilation.  The remaining licenses (BSD, MIT, and Boost license, as far as I know) are all permissive.  Therefore, this build should be free of copyleft restrictions.

A few issues:
This build uses custom implementations of many number-theoretic functions that GMP provides but Boost.multiprecision does not (see mp_class.h and mp_boost.cpp).  Some of them are quite slow and should be updated in the future.
When INTEGER_CLASS=boostmp, the compiler's level 3 optimizations (-O3) are automatically turned off in Release mode.  Otherwise, some tests segfault.  This may be a result of undefined behavior somewhere this pull request, elsewhere in Symengine, or in Boost.multiprecision itself.  I haven't been able to track it down yet.
Boost's random number generation types don't line up perfectly with GMP's. Also, GMP types like gmp_randstate_t were never wrapped, and other authors used such types directly at times. With boostmp, all code that requires random number generation has either been removed vid ifdef’s or reimplemented via ifdef's.  It might be cleaner to create a uniform interface for random number generation (mp_randstate, etc) and wrap both GMP and Boost's random number generation into that interface.
Boost.multiprecision does not provide non-const access to the numerator and denominator of its rational class.  I have worked around the few instances in which SymEngine relies on such access with ifdef's.  We might consider removing the non-const versions of get_num and get_den for all builds if they are seldom used.

isuruf commented:
@ivme, good job. :100: 
I tried this with boost 1.58 and there were a few compile errors. One error is that `make_unsigned` is not defined for the boostmp int which is needed for the random numbers. Error is here, https://gist.github.com/isuruf/0f7b6612965a6e93034407923d56ead5

After specializing `make_unsigned_imp` and a few other fixes, tests pass with `-O3`. I'll push the fixes soon.

Thanks for doing this once again.

ivme commented:
Thanks, @isuruf!

I pulled your commits into my local repo and modified CMakeLists to build with -O3.  As before, the build succeeds but I get errors in the tests.  I wonder if this is a difference between boost 1.58 and 1.62.  Configuration and ctest results here:
https://gist.github.com/ivme/6598812322c2f71859f75952fc566a4b

isuruf commented:
@ivme, can you post the output of `ctest --output-on-failure` ? I'll try 1.62 as well.

ivme commented:
@isuruf I've updated the [gist](https://gist.github.com/ivme/6598812322c2f71859f75952fc566a4b) with results of `ctest --output-on-failure`.  System is OS X 10.11.6, Apple LLVM version 8.0.0 (clang-800.0.38)

isuruf commented:
@ivme, all of the tests are related to polynomials. I've tried boost 1.62 on Ubuntu 16.04 and it works fine.

ivme commented:
@isuruf When I switched to Linux and compiled with gcc and -O3 I got no errors, so I tried compiling with gcc instead of clang on OS X -- also no errors.  This might be a clang issue specifically.  I've updated the condition to remove the -O3 flag only if the compiler is clang.  Still only a temporary fix.

It seems the problem has something to do with copying the std::map that provides the backing for ODictWrapper when ODictWrapper's assignment operator is called.  If you have clang handy and you're interested in investigating, there are some test cases that demonstrate the fundamental issue in my [debug](https://github.com/ivme/symengine/tree/debug) branch, under symengine/tests/debugging.

isuruf commented:
Hmm, just tried with clang 3.8 on Linux and it works. No segfaults.

certik commented:
@ivme thanks a lot for this contribution! It's been long on my todo list to implement this, so that we can build symengine with permissive licenses only, if so desired. This will be very useful to a lot of people.

isuruf commented:
@ivme, can you try clang from homebrew?

ivme commented:
@isuruf, I installed clang 3.8 via homebrew and built symengine with -O3 flag.  still failing tests.

ivme commented:
Installed clang-3.8 via apt-get on ubuntu 16.04 and built SymEngine with -O3.  All tests pass.

certik commented:
@ivme looks like all tests pass. Is there some thing that still has to be fixed in this PR?

isuruf commented:
We should add a test. I'm waiting on #1119

isuruf commented:
@ivme, any idea what's causing this error, https://travis-ci.org/symengine/symengine/jobs/176270177#L575-L607 ?

isuruf commented:
Tests pass now. @certik, a final review?

certik commented:
I don't like that there is `#if SYMENGINE_INTEGER_CLASS != SYMENGINE_BOOSTMP` at quite a few places in the code. But it's not at that many places, so I am ok with merging this as it is.

Usually there is a way to rework the integer class so that all these ifdef statements go in it, and outside of it, i.e. the users of it (the rest of symengine) just use the exact same code for both gmp or boost or any other implementation of an integer. But we can do this in future PRs.

My last question is if either of you is willing to tidy the history a bit, and create a few nice logical commits. One would do `git reset --soft HASH` where HASH is the current master and then just add the changes using "git add -p" or just "git add file" and commit when it makes sense.

ivme commented:
Yes, I agree about all the `#if SYMENGINE_INTEGER_CLASS != SYMENGINE_BOOSTMP` markings.  I considered making more changes to the multiprecision integer interface (e.g. wrapping GMP and Boost's random number functionality into a unified interface) to eliminate some of these `#if`'s, but I thought I might check with the rest of the development team first, since this is my first contribution.  I'd be happy to do some cleanup in this or a future PR.  I can do some logical commits also, but I won't have much time in the next week or so, so if Isuru would like to do it, by all means.

isuruf commented on symengine/rational.h:
```diff
@@ -20,8 +20,17 @@ class Rational : public Number
     IMPLEMENT_TYPEID(RATIONAL)
-    //! Constructor of Rational class
+//! Constructor of Rational class
+#if SYMENGINE_INTEGER_CLASS == SYMENGINE_BOOSTMP
```
These constructors ca be enabled for all types, right?

ivme replied:
meant to ask about that. The only constructor for `SymEngine::Rational` is `Rational(rational_class i)`.  Likewise `SymEngine::Integer` has only the constructor `Integer(integer_class i)`.  Because neither parameter is passed by reference, copies are made in both cases.  The `Rational` constructor then initializes the private member `i` using `i{i}`, again copying `i`.  The `Integer` constructor initializes the private member `i` with `i{std::move(i)}`, avoiding the second copy.  The difference strikes me as purposeful, but I don't understand the reason.

It seems more logical to me to have two constructors for each class, much like copy and move constructors, but from rational_class and integer_class respectively:
`Rational(const rational_class &_i)`
`Rational(rational_class &&_i)`
`Integer(const integer_class &_i)`
`Integer(integer_class &&_i)`

isuruf replied:
Yes, we should have all 4 methods. This was probably there because we couldn't figure out how to enable move semantics of `mpz_class` without enabling other C++11 features. (SymEngine supported gcc 4.6 back then).  `i{std::move(i)}` is a newer addition, I think.
Anyways, we should have all 4 methods you mentioned.

isuruf commented:
There are 2 major differences that require `#ifdef`s.

>  It might be cleaner to create a uniform interface for random number generation (mp_randstate, etc) and wrap both GMP and Boost's random number generation into that interface.

This is easy to fix, but we can do it another PR as well. I might be able to look at this in the weekend.

> Boost.multiprecision does not provide non-const access to the numerator and denominator of its rational class

This cannot be solved easily, so these `#ifdef`s has to be there.

certik commented:
I suggest we merge as it is, and polish in subsequent PRs.

Let's tidy the history though. If neither of you can do it, I can have a look myself. @isuruf let me know on Monday, and I can have a look next week.

isuruf commented:
@ivme, @certik, I've rebased this on top of master and then squashed some commits. Also reduced some special casing by using @ivme 's new `mp_randstate` class.

certik commented:
@ivme, @isuruf thanks for the work. I merged it, as I think it's more important to have this in, rather than a super clean history. I don't have time right now to squash the history more and rather than getting it stalled, let's get it in.
